// web-api/lib/createRequest.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

exports.default = createRequest;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createRequest(method, source, url) {
  var options = arguments.length <= 3 || arguments[3] === undefined ? {} : arguments[3];
  var reqOptions = arguments.length <= 4 || arguments[4] === undefined ? {} : arguments[4];
  var _options$queryParams = options.queryParams;
  var queryParams = _options$queryParams === undefined ? {} : _options$queryParams;
  var _options$queryParamMa = options.queryParamMappings;
  var queryParamMappings = _options$queryParamMa === undefined ? {} : _options$queryParamMa;
  var _options$params = options.params;
  var params = _options$params === undefined ? {} : _options$params;


  var fullUrl = source.baseURL + url;

  params.method = method.toUpperCase();
  params.headers = (0, _extends3.default)({}, params.headers, source.globalHeaders);

  return {
    params: params,
    queryParams: queryParams,
    queryParamMappings: queryParamMappings,
    reqOptions: reqOptions,
    url: fullUrl,
    fetchUrl: ''
  };
}
